pkg/repository/maintenance/maintenance.go

```diff
@@ -0,0 +1,494 @@
+	repo *velerov1api.BackupRepository,
+) (*JobConfigs, error) {
+	var cm v1.ConfigMap
+	if err := client.Get(
+		ctx,
+		types.NamespacedName{
+			Namespace: veleroNamespace,
+			Name:      repoMaintenanceJobConfig,
+		},
+		&cm,
+	); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, nil
+		} else {
+			return nil, errors.Wrapf(
+				err,
+				"fail to get repo maintenance job configs %s", repoMaintenanceJobConfig)
+		}
+	}
+
+	if cm.Data == nil {
+		return nil, errors.Errorf("data is not available in config map %s", repoMaintenanceJobConfig)
+	}
+
+	// Generate the BackupRepository key.
+	// If using the BackupRepository name as the is more intuitive,
+	// but the BackupRepository generation is dynamic. We cannot assume
+	// they are ready when installing Velero.
+	// Instead we use the volume source namespace, BSL name, and the uploader
+	// type to represent the BackupRepository. The combination of those three
+	// keys can identify a unique BackupRepository.
+	repoJobConfigKey := repo.Spec.VolumeNamespace + "-" +
+		repo.Spec.BackupStorageLocation + "-" + repo.Spec.RepositoryType
+
+	var result *JobConfigs
+	if _, ok := cm.Data[repoJobConfigKey]; ok {
+		logger.Debugf("Find the repo maintenance config %s for repo %s", repoJobConfigKey, repo.Name)
+		result = new(JobConfigs)
+		if err := json.Unmarshal([]byte(cm.Data[repoJobConfigKey]), result); err != nil {
+			return nil, errors.Wrapf(
+				err,
+				"fail to unmarshal configs from %s's key %s",
+				repoMaintenanceJobConfig,
+				repoJobConfigKey)
+		}
+	}
+
+	if _, ok := cm.Data[GlobalKeyForRepoMaintenanceJobCM]; ok {
+		logger.Debugf("Find the global repo maintenance config for repo %s", repo.Name)
+
+		if result == nil {
+			result = new(JobConfigs)
+		}
+
+		globalResult := new(JobConfigs)
+
+		if err := json.Unmarshal([]byte(cm.Data[GlobalKeyForRepoMaintenanceJobCM]), globalResult); err != nil {
+			return nil, errors.Wrapf(
+				err,
+				"fail to unmarshal configs from %s's key %s",
+				repoMaintenanceJobConfig,
+				GlobalKeyForRepoMaintenanceJobCM)
+		}
+
+		if result.PodResources == nil && globalResult.PodResources != nil {
+			result.PodResources = globalResult.PodResources
+		}
+
+		if len(result.LoadAffinities) == 0 {
+			result.LoadAffinities = globalResult.LoadAffinities
+		}
+	}
+
+	return result, nil
+}
+
+// WaitJobComplete waits the completion of the specified maintenance job and return the BackupRepositoryMaintenanceStatus
+func WaitJobComplete(cli client.Client, ctx context.Context, jobName, ns string, logger logrus.FieldLogger) (velerov1api.BackupRepositoryMaintenanceStatus, error) {
+	log := logger.WithField("job name", jobName)
+
+	maintenanceJob, err := waitForJobComplete(ctx, cli, ns, jobName)
+	if err != nil {
+		return velerov1api.BackupRepositoryMaintenanceStatus{}, errors.Wrap(err, "error to wait for maintenance job complete")
+	}
+
+	log.Info("Maintenance repo complete")
+
+	result, err := getResultFromJob(cli, maintenanceJob)
```

Comment:
trivial: Now it seems the result will be empty string if the job is successful, maybe we should update the maintennance.go under cli package to make sure it returns a message when it's successful.

Reply:
It is expected and as the old logic.
For the old code, the result is not empty when and only when maintenance fails. And the old logic use `result != ""` to judge it is a failed maintenance job.
For the new code, we only use the error returned by `getResultFromJob` to tell if it is a failed job. But the result is as is --- it is not empty when and only when maintenance fails.